src/CoreWCF.AzureQueueStorage/src/CoreWCF/Channels/AzureQueueStorageQueueTransport.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+using CoreWCF.Configuration;
+using CoreWCF.Queue;
+using CoreWCF.Queue.Common;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace CoreWCF.Channels
+{
+    public class AzureQueueStorageQueueTransport : IQueueTransport
+    {
+        private readonly string _connectionString;
+        private readonly string _queueName;
+        private readonly string _deadLetterQueueConnectionString;
+        private readonly string _deadLetterQueueName;
+        private readonly QueueClient _queueClient;
+        private readonly QueueClient _deadLetterQueueClient;
+        private readonly TimeSpan _queueReceiveTimeOut;
+        private readonly TimeSpan _receiveMessagevisibilityTimeout;
+        private readonly ILogger<AzureQueueStorageQueueTransport> _logger;
+
+        public AzureQueueStorageQueueTransport(IServiceDispatcher serviceDispatcher, IServiceProvider serviceProvider)
+        {
+            _queueClient = new QueueClient(_connectionString, _queueName);
+            _deadLetterQueueClient = new QueueClient(_deadLetterQueueConnectionString, _deadLetterQueueName);
+            _queueReceiveTimeOut = serviceDispatcher.Binding.ReceiveTimeout;
+            _logger = serviceProvider.GetRequiredService<ILogger<AzureQueueStorageQueueTransport>>();
+        }
+
+        public int ConcurrencyLevel => 1;
+
+        public async ValueTask<QueueMessageContext> ReceiveQueueMessageContextAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            _logger.LogInformation("Receiving message from Azure queue storage");
+
+            QueueMessage queueMessage = await _queueClient.ReceiveMessageAsync(_receiveMessagevisibilityTimeout, cancellationToken);
+            if(queueMessage == null)
+            {
+                return null;
+            }
+            _queueClient.DeleteMessage(queueMessage.MessageId, queueMessage.PopReceipt);
+            var reader = PipeReader.Create(new ReadOnlySequence<byte>(queueMessage.Body.ToMemory()));
```

Comment:
Where are you passing a visibilityTimeout but then immediately deleting the message? My understanding is this is how long until the message gets made available again. If so, you should hold off deleting the message until it's completed. There's a class (made public in the PR by Jon from AWS) called ReceiveContext that you can add to the Message which has a OnCompleteAsync and OnAbortAsync method. In OnCompleteAsync, you would delete the message. In the OnAbortAsync method, you would push the message to the dead letter queue.

Reply:
I was deleting the message after reading it so as to remove it from the queue but as  you said it makes sense to put in the OnCompleteAsync method. but the PR you mentioned is not merged into the repository yet. i see it still open. So i am thinking i should wait for it to be pushed or do you think i should merge the changes anyway?

Reply:
as discussed, i will make the changes once the PR is merged